### PR TITLE
Implement qualityLabels

### PR DESCRIPTION
**Pull request by johnBartos**
RFC on the quality-labels utils. Not sure how clear/clean the code is - I developed it with TDD in mind.

https://github.com/jwplayer/jwplayer-commercial/pull/2920 (2/2) 
JW7-2266

### CODE REVIEW

File: test/unit/quality-labels.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+       assert.equal(actual, expected);
+    });
+
+    test('retuns null when no custom labels are given', function (assert) {
+        var expected = null;
+        var actual = qualityLabels.getCustomLabel(null, 4000000);
+        assert.equal(actual, expected);
+    });
+
+    test('retuns null when no custom labels are empty', function (assert) {
```
Comment (egreaves) [inline test/unit/quality-labels.js]: remove "no" from the test description

File: test/unit/quality-labels.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        var expected = null;
+        var actual = qualityLabels.getCustomLabel([], 4000000);
+        assert.equal(actual, expected);
+    });
+
+    test('retuns null when bandwidth is undefined', function (assert) {
+        var expected = null;
+        var actual = qualityLabels.getCustomLabel([], undefined);
```
Comment (egreaves) [inline test/unit/quality-labels.js]: Though this passes, it's misleading because your bandwidths array is also empty, which could indicate a false positive. Consider expanding this to test `bandwidth === undefined` with both an empty array and an array containing at least 1 bandwidth. Alternatively, you could just test it with a non-empty bandwidths array since the test above covers the case where `bandwidths.length === 0`.

File: src/js/api/config.js [hunk 3]
```diff
@@ -126,6 +126,8 @@ define([
             config.playlist = [ obj ];
         }
 
+        config.qualityLabels = config.qualityLabels || config.hlslabels;
```
Comment (hghazzi) [inline src/js/api/config.js]: We can verify with @esboyd, but I believe all config names you be entirely lowercase `qualitylabels`.

Comment (johnBartos) [inline src/js/api/config.js]: `hlslabels` isn't camelcase

File: test/unit/quality-labels-test.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    var test = QUnit.test.bind(QUnit);
+
+    var customLabels  = {
+        1000: 'low',
+        2000: 'medium',
+        3000: 'high',
+    };
+
+    var bandwidths = _.keys(customLabels);
+
```
Comment (hghazzi) [inline test/unit/quality-labels-test.js]: Can you add `Qunit.module` before each method thats being tested? Its a useful separator, just like how we would use `describe` in mocha.

File: test/unit/quality-labels-test.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        var actual = qualityLabels.createLabel(360, undefined, false);
+        assert.equal(actual, expected);
+    });
+
+    test('creates a label using bandwidth when height is not available', function (assert) {
+        var expected = '1337 kbps';
+        var actual = qualityLabels.createLabel(undefined, 1337000,  false);
+        assert.equal(actual, expected);
+    })
+
```
Comment (hghazzi) [inline test/unit/quality-labels-test.js]: Add a test with args `undefined, 1337000, true` to verify that parentheses are not added.

File: test/unit/quality-labels-test.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        var expected = null;
+        var actual = qualityLabels.getCustomLabel(customLabels, undefined);
+        assert.equal(actual, expected);
+    });
+
+    test('retuns null when no bandwidths are given', function (assert) {
+        var expected = null;
+        var actual = qualityLabels.findClosestBandwidth([], 4000);
+        assert.equal(actual, expected);
+    });
```
Comment (hghazzi) [inline test/unit/quality-labels-test.js]: Same

File: test/unit/quality-labels-test.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        var expected = null;
+        var actual = qualityLabels.getCustomLabel(null, 4000000);
+        assert.equal(actual, expected);
+    });
+
+    test('retuns null when custom labels are empty', function (assert) {
+        var expected = null;
+        var actual = qualityLabels.getCustomLabel([], 4000000);
+        assert.equal(actual, expected);
+    });
```
Comment (hghazzi) [inline test/unit/quality-labels-test.js]: Edge case, but add a test where you pass in a non array value, since that would pass a guard that is only looking for a falsey value.

File: src/js/providers/flash.js [hunk 8]
```diff
@@ -125,11 +74,8 @@ define([
...
                 // Set original index
                 level.index = i;
-                // Translate sources into quality levels, assigning custom levels if present
-                if (_customLabels && level.bitrate) {
-                    // get label with nearest rate match
-                    var sourceKbps = Math.round(level.bitrate / 1000);
-                    level.label = _getNearestCustomLabel(sourceKbps);
+                if (level.label !== 'Auto') {
+                    level.label = utils.generateLabel(level, _playerConfig.qualityLabels);
                 }
```
Comment (hghazzi) [inline src/js/providers/flash.js]: Do we need to handle the case where two labels could end up having the same label if they are both closest to the same bandwidth?

File: src/js/api/config.js (hunk 3 above)
Comment (johnBartos) [inline src/js/api/config.js]: wontfix, see `minDvrWindow` and `timeSliderAbove`

File: test/unit/quality-labels-test.js (hunk 7 above)
Comment (johnBartos) [inline test/unit/quality-labels-test.js]: qualityLabels should actually be {} in this case

File: src/js/providers/flash.js (hunk 8 above)
Comment (johnBartos) [inline src/js/providers/flash.js]: They will both have the same label

File: test/unit/quality-labels-test.js (hunk 4 above)
Comment (johnBartos) [inline test/unit/quality-labels-test.js]: done

File: test/unit/quality-labels-test.js (hunk 6 above)
Comment (johnBartos) [inline test/unit/quality-labels-test.js]: done
